### PR TITLE
add http client factory support https://github.com/justeat/ZendeskApi…

### PR DESCRIPTION
**Pull request by jonathanstowell**
…Client/issues/141

### CODE REVIEW

File: src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -1,5 +1,8 @@
+using System;
+using System.Net.Http;
 using Microsoft.Extensions.DependencyInjection;
 using ZendeskApi.Client.Options;
+#pragma warning disable 618
```
Comment (LoadData) [inline src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs]: Why's this added?

File: src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -23,6 +26,30 @@ string token
             return services;
         }
 
+        public static IServiceCollection AddZendeskClientWithHttpClientFactory(this IServiceCollection services,
+            string endpointUri,
+            string username,
+            string token,
+            Action<HttpClient> configureClient = null
+        )
```
Comment (LoadData) [inline src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs]: Can we move this up a line?

File: src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -38,5 +65,27 @@ string oAuthToken
 
             return services;
         }
+
+        public static IServiceCollection AddZendeskClientWithHttpClientFactory(this IServiceCollection services,
+            string endpointUri,
+            string oAuthToken,
+            Action<HttpClient> configureClient = null
+        )
```
Comment (LoadData) [inline src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs]: Can we move this up a line?

File: src/ZendeskApi.Client/ZendeskApiClientFactory.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            IHttpClientFactory httpClientFactory)
+        {
+            _options = options.Value;
+            _httpClientFactory = httpClientFactory;
+        }
+
+        public HttpClient CreateClient(string resource = null)
+        {
+            var formattedResource = resource?
+                .Trim('/');
```
Comment (LoadData) [inline src/ZendeskApi.Client/ZendeskApiClientFactory.cs]: Can we put the `Trim` call onto the line above too?

File: test/ZendeskApi.Client.IntegrationTests/Factories/ZendeskClientFactory.cs [hunk 5]
```diff
@@ -1,6 +1,7 @@
 using Microsoft.Extensions.Options;
 using ZendeskApi.Client.IntegrationTests.Settings;
 using ZendeskApi.Client.Options;
+#pragma warning disable 618
```
Comment (LoadData) [inline test/ZendeskApi.Client.IntegrationTests/Factories/ZendeskClientFactory.cs]: Why is this included?

File: test/ZendeskApi.Client.Tests/ZendeskApiClientTests.cs [hunk 6]
```diff
@@ -4,6 +4,7 @@
 using Microsoft.Extensions.Options;
 using Xunit;
 using ZendeskApi.Client.Options;
+#pragma warning disable 618
```
Comment (LoadData) [inline test/ZendeskApi.Client.Tests/ZendeskApiClientTests.cs]: Why's this included?

Comment (jonathanstowell) [reply test/ZendeskApi.Client.Tests/ZendeskApiClientTests.cs]: Because the old ZendeskApiClient is marked as Obsolete and produces a warning which we don't want as we'll remove it eventually. We still want to keep the tests.

File: test/ZendeskApi.Client.IntegrationTests/Factories/ZendeskClientFactory.cs (hunk 5 above)
Comment (jonathanstowell) [reply test/ZendeskApi.Client.IntegrationTests/Factories/ZendeskClientFactory.cs]: Because the old ZendeskApiClient is marked as Obsolete and produces a warning which we don't want as we'll remove it eventually. We still want to keep the tests.

File: src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs [hunk 7]
```diff
@@ -1,5 +1,8 @@
+using System;
+using System.Net.Http;
 using Microsoft.Extensions.DependencyInjection;
 using ZendeskApi.Client.Options;
+#pragma warning disable 618
 
 namespace ZendeskApi.Client.Extensions
```
Comment (Jetski5822) [inline src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs]: Can we remove the Extensions namespace please?

File: src/ZendeskApi.Client/ZendeskApiClientFactory.cs (hunk 4 above)
Comment (jonathanstowell) [reply src/ZendeskApi.Client/ZendeskApiClientFactory.cs]: I've been formatting to have one operation per line as its easier to see what's happening when things change via a diff tool.

File: src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs (hunk 1 above)
Comment (jonathanstowell) [reply src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs]: Because the old ZendeskApiClient is marked as Obsolete and produces a warning which we don't want as we'll remove it eventually.

File: src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs (hunk 7 above)
Comment (jonathanstowell) [reply src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs]: Do we want to do that, it will break consumers implementations if they're already using these extensions? It's been in this namespace for 2 years https://github.com/justeat/ZendeskApiClient/commit/30cfb9a10c39f886853c6af7d63863593eadaee5#diff-7d9f67410130ec9e68e6f5eb87ae63bbR4

Comment (Jetski5822) [reply src/ZendeskApi.Client/Extensions/ServiceCollectionExtensions.cs]: Well I always considered it a bit of a bad practice to use that namespace. 

But yeah - I guess if you don't want to break exciting clients, maybe leave as is - I should have paid more attention when it was changed 2 years ago LOL

**Comment by Jetski5822** [comment]
BTW - great idea to move to IHttpClientFactory. Nice one!
